src/core/solvers/ipopt/multiple-shooting-nlp.cpp

```diff
@@ -0,0 +1,589 @@
+      nvar_(T_ * (ndx_ + nu_) + ndx_)  // Multiple shooting, states and controls
+{
+    xs_.resize(T_ + 1);
+    us_.resize(T_);
+
+    for (size_t i = 0; i < T_; i++) {
+        xs_[i] = state_->zero();
+        us_[i] = Eigen::VectorXd::Zero(nu_);
+    }
+    xs_[T_] = xs_[0];
+
+    data_.x      = state_->zero();
+    data_.xnext  = state_->zero();
+    data_.dx     = Eigen::VectorXd::Zero(ndx_);
+    data_.dxnext = Eigen::VectorXd::Zero(ndx_);
+    data_.x_diff = Eigen::VectorXd::Zero(ndx_);
+
+    data_.control = Eigen::VectorXd::Zero(nu_);
+
+    data_.Jsum_x      = Eigen::MatrixXd::Zero(ndx_, ndx_);
+    data_.Jsum_dx     = Eigen::MatrixXd::Zero(ndx_, ndx_);
+    data_.Jsum_xnext  = Eigen::MatrixXd::Zero(ndx_, ndx_);
+    data_.Jsum_dxnext = Eigen::MatrixXd::Zero(ndx_, ndx_);
+
+    data_.Jdiff_xnext = Eigen::MatrixXd::Zero(ndx_, ndx_);
+    data_.Jdiff_x     = Eigen::MatrixXd::Zero(ndx_, ndx_);
+
+    data_.Jg_dx     = Eigen::MatrixXd::Zero(ndx_, ndx_);
+    data_.Jg_dxnext = Eigen::MatrixXd::Zero(ndx_, ndx_);
+    data_.Jg_u      = Eigen::MatrixXd::Zero(ndx_, ndx_);
+
+    data_.Ldx   = Eigen::VectorXd::Zero(ndx_);
+    data_.Ldxdx = Eigen::MatrixXd::Zero(ndx_, ndx_);
+    data_.Ldxu  = Eigen::MatrixXd::Zero(ndx_, nu_);
+}
+
+MultipleShootingNlp::~MultipleShootingNlp() {}
+
+bool MultipleShootingNlp::get_nlp_info(Ipopt::Index   &n,
+                                       Ipopt::Index   &m,
+                                       Ipopt::Index   &nnz_jac_g,
+                                       Ipopt::Index   &nnz_h_lag,
+                                       IndexStyleEnum &index_style)
+{
+    n = nvar_;
+
+    m = nconst_;
+
+    // Jacobian nonzeros for dynamic constraints
+    nnz_jac_g = T_ * ndx_ * (2 * ndx_ + nu_);
+
+    // Jacobian nonzeros for initial condition
+    nnz_jac_g += ndx_;
+
+    // Hessian is only affected by costs
+    // Running Costs
+    std::size_t nonzero = 0;
+    for (size_t i = 1; i <= (ndx_ + nu_); i++) {
+        nonzero += i;
+    }
+    nnz_h_lag = T_ * nonzero;
+
+    // Terminal Costs
+    nonzero = 0;
+    for (size_t i = 1; i <= ndx_; i++) {
+        nonzero += i;
+    }
+    nnz_h_lag += nonzero;
+
+    // use the C style indexing (0-based)
+    index_style = Ipopt::TNLP::C_STYLE;
+
+    return true;
+}
+// [TNLP_get_nlp_info]
+
+// [TNLP_get_bounds_info]
+// returns the variable bounds
+bool MultipleShootingNlp::get_bounds_info(Ipopt::Index   n,
+                                          Ipopt::Number *x_l,
+                                          Ipopt::Number *x_u,
+                                          Ipopt::Index   m,
+                                          Ipopt::Number *g_l,
+                                          Ipopt::Number *g_u)
+{
+    // here, the n and m we gave IPOPT in get_nlp_info are passed back to us.
+    // If desired, we could assert to make sure they are what we think they are.
+    assert(n == nvar_);
+    assert(m == nconst_);
```

Comment:
Let's use ``assert_pretty` as it triggers issues in the Python code as well (without the need to compile in Debug mode)